### ISSUE TITLE
Is it possible to code the header at the end?

### ISSUE THREAD

**Issue by pkbehera** [seleznevae/libfort#28]
I am dynamically populating a table in a function which returns ta string by calling `to_string()` in the end. My logic is such that I may end up in not adding any data at all to the table. In such cases I don't want to add any header to the table, so that `to_string()` returns an empty string.

Hence a feature to add the header at the end after knowing the number of rows (or whenever we wish) will be helpful. Currently I use workarounds to return an empty string from my function, as adding the header afterwards adds it after all the data added till then.

**Comment by seleznevae** [comment]
Hi!
I think it is not possible to do it at the moment.
At the moment it is possible to change current cell where content will be inserted. In current implementation after that new content will overwrite existing one. 
As far as I understand you want to get another behavior - new inserted content should be in a brand new row and the old rows should just go below.
So as a generalization: you want somehow change (or be able to change) current strategy "overwrite old content with the new one" with the new strategy "insert new content, old content preserve" for content adding functions. Do I understand your problem correctly?

**Comment by pkbehera** [comment]
yes the snippet you posted is what I want to do.. For me as an user,  the header line should always be the at the top of the table; irrespective of the time when it is added/coded.

```
int main()
{
    fort::char_table table;
    table << fort::header << "N"
          << "Driver"
          << "Time"
          << "Avg Speed" << fort::endr;
    table << "1"
          << "Ricciardo"
          << "1:25.945"
          << "47.362" << fort::endr;
    table << "2"
          << "Hamilton"
          << "1:26.373"
          << "35.02" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```
please note that to add the header we need to do insert a **special attribute** to the able i.e. `table << fort::header`

**So the table knows which is the header row. So why must we add the header the very beginning?**  Why can we just add it whenever we want to?

**Comment by pkbehera** [comment]
In my view all these code samples should produce the same output, **but they don't:**
```
int main()
{
    fort::char_table table;
    table << fort::header << "N"
          << "Driver"
          << "Time"
          << "Avg Speed" << fort::endr;
    table << "1"
          << "Ricciardo"
          << "1:25.945"
          << "47.362" << fort::endr;
    table << "2"
          << "Hamilton"
          << "1:26.373"
          << "35.02" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```
```
int main()
{
    fort::char_table table;
    table << "1"
          << "Ricciardo"
          << "1:25.945"
          << "47.362" << fort::endr;
    table << "2"
          << "Hamilton"
          << "1:26.373"
          << "35.02" << fort::endr;
    table << fort::header << "N"
          << "Driver"
          << "Time"
          << "Avg Speed" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```
```
int main()
{
    fort::char_table table;
    table << "1"
          << "Ricciardo"
          << "1:25.945"
          << "47.362" << fort::endr;
    table << fort::header << "N"
          << "Driver"
          << "Time"
          << "Avg Speed" << fort::endr;
    table << "2"
          << "Hamilton"
          << "1:26.373"
          << "35.02" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```

**Comment by pkbehera** [comment]
btw what does `<< fort::header` do? Apparently its **just** equivalent to `<< fort::endr << fort::separator`

**Comment by seleznevae** [comment]
I got your idea. It doesn't work the way you thing because in *libfort* there is no a distinct entity such as header. All rows (ordinary rows and headers) are pretty much the same. In *libfort* header is just a property of the row. Rows that have this property just have another border and that's all. 
Approach when we treat *headers* as some kind of special entity is simply not flexible (e.g. you can't have headers at the bottom, multiple headers etc.). 
`fort::header` is just setting property for the row (it equivalent to `ft_set_cell_prop(table_, FT_CUR_ROW, FT_ANY_ROW, FT_CPROP_ROW_TYPE, FT_ROW_HEADER);`)

**Comment by pkbehera** [comment]
well, spreadsheet applications for example treat headers as special entities, and allow to customize the header in many ways, for example put it in every page in a printout. :)

**Comment by seleznevae** [comment]
Yes, I understand what you mean but as I said this approach with treating headers and ordinary rows differently is more complicated and also I believe current approach is more flexible. All that logic with some special behavior can be added externally in some kind of a wrapper(maybe I am wrong). So I prefer to keep things simple and implement more basic functionality.
Returning to you initial question. I can suggest adding new property to the tables (lets call it **CONTENT_ADDING_STRATEGY** (with 2 possible values **REPLACE**(default one), **INSERT**). I think this approach is rather flexible. Also it will fit well for some other scenarios.

In this case your code will look like this.
```
// in c++ api
std::string some_function()
{
  fort::char_table tbl;
  tbl.set_adding_strategy(fort::strategy::INSERT);
  // populate tbl with data

  // add header if some data were really added
  if (!tbl.is_empty()) {
     // note that you have to explicitly set current cell here
     tbl.set_cur_cell(0, 0); 
     // you explicitly add header in place of the 0th row all other rows will be shifted
     tbl << fort::header << "hdr1" << "hdr2" << "hdr3";
  }
  return tbl.to_string();
}

// Also if you don't like to set this property for each table
// you can set it one time during some initialization
fort::default_props().set_adding_strategy(fort::strategy::INSERT);
```
Does it suit you? Or maybe you don't like it?

**Comment by seleznevae** [comment]
Hi. I implemented *adding_strategies* and merged branch to `develop`.
Now it is possible to add rows (and headers) at any time to any place in the table. 
Here is a simple example:

```c++
int main()
{
    fort::char_table table;
    table.set_adding_strategy(fort::add_strategy::insert);

    // Fill table with data
    table << "1" << "2" << fort::endr;
    table << "3" << "4" << fort::endr;

    // explicitly add header as the top row
    table.set_cur_cell(0, 0);
    table << fort::header
          << "hdr1" << "hdr2" << fort::endr;

    std::cout << table.to_string() << std::endl;
}
```
Output:
```
+------+------+
| hdr1 | hdr2 |
+------+------+
| 1    | 2    |
| 3    | 4    |
+------+------+
```
The syntax is rather verbose but it is simple and straightforward. You have to explicitly specify `adding_strategy` as `replace`. After that when you want to insert some data inside the table you explicitly set current cell with `set_cur_cell` and then use ordinary functions to fill the cells. 

If this solution doesn't solve your problems fill free to write here your suggestions or problems that you encounter.

**Comment by pkbehera** [comment]
thanks, so now we can code the header either at the start or at the very end. I tried the following

```
int main()
{
    fort::char_table table;
    table.set_adding_strategy(fort::add_strategy::insert);

    // Fill table with data
    table << "1"
          << "2" << fort::endr;
    // explicitly add header as the top row
    table.set_cur_cell(0, 0);
    table << fort::header << "hdr1"
          << "hdr2" << fort::endr;
    table << "3"
          << "4" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```
and this was the output
```
+------+------+
| hdr1 | hdr2 |
+------+------+
| 3    | 4    |
| 1    | 2    |
+------+------+
```
adding more data after adding the header gives counter-intuitive results.

**Comment by pkbehera** [comment]
Pressed the wrong button and the issue got closed...

**Comment by seleznevae** [comment]
That's expected behavior.
As I already said above there is nothing special with headers. They are ordinary rows. That's library design principle. 
Functions `set_adding_strategy` and `set_cur_cell` also ordinary functions. They don't do any specific for headers. They designed in a general way so that they affect all inserted cells(either in headers or in ordinary rows). There is no reason to limit the functionality of adding cells to random place only for headers. Therefore when you specify that you will insert data to the beginning of the table library will add there all rows (headers and ordinary rows). It's a library user responsibility to specify where to place data. So in your last example if you want to add `3` and `4` to the end you have to explicitly specify it with `table.set_curl_cell(2, 0);` before pushing these data.
However, maybe what you want can be implemented with a little bit different API. Something like that:
```c++
int main()
{
    fort::char_table table;
    table.set_adding_strategy(fort::add_strategy::insert);

    // Fill table with data
    table << "1" << "2" << fort::endr;

    // Note that I explicitly specify row where I insert the header
    table[0] << fort::header << "hdr1"  << "hdr2" << fort::endr;

    table << "3" << "4" << fort::endr;
    std::cout << table.to_string() << std::endl;
}
```
So the idea of this API is that we can  apply modifiing operations not on the table as a hole but on the rows and that in this case this operations won't affect position of the current cell. I need to think about this possibility thoroughly and also how it will look with current  modifying operation on distinct cells (like `tbl[2][3] = "some_data";` ) which affect position of the current cell. So I am not sure if it will be ok to implement it. Need to think about it anyway.

**Comment by brlcad** [comment]
Wouldn't a header at the end typically be called a footer? :)

Since it is just decorative, perhaps renaming it as a cell/row border bolding or callout property?  Something like:  tables[123] << fort::emphasize << "Total:" << "123" << fort::endr;

**Comment by pkbehera** [comment]
appending the table at the end will be a footer, but **pre-pending** the table at the end will still be a header! :)
